### PR TITLE
Usersystem + Database

### PR DESCRIPTION
**Pull request by xWafl**
Added a database and users system, originally was just gonna be a users system but it expanded

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -9,30 +9,64 @@ const moment = require('moment');
 // const io = require('socket.io').listen(http);
 const port = process.env.PORT || 5000;
 
-// const server = app()
-    // .use((req, res) => res.sendFile(INDEX, { root: __dirname }))
-    // .listen(port, () => console.log(`Listening on ${port}`));
```
Comment (Mateiadrielrafael) [inline index.ts]: Dont keep ghost code

File: index.ts [hunk 2]
```diff
@@ -9,30 +9,64 @@ const moment = require('moment');
...
 
-const wss = new Server({ server });
+const {setWsHeartbeat} = require("ws-heartbeat/server");
+const {Server} = require('ws');
 
-class User {
-    constructor(public username: string, public usernum: number, public currentChannel: number, public id: number, public messages: Array<number>){}
+interface User {
+    username: string,
+    userNum: number,
```
Comment (Mateiadrielrafael) [inline index.ts]: Idk, it kinda looks bad to have the n in username be lowercase and he one in useNum to be uppercase, but maybe it's just me

File: index.ts [hunk 3]
```diff
@@ -9,30 +9,64 @@ const moment = require('moment');
...
+    currentChannel: number,
+    id: number,
+    messages: Array<number>
 }
 
-class Message {
-    constructor(public user: User, public utctime: string, public date: string, public message: string, public channel: string, public id: number, public active: boolean){}
+interface Message {
+    user: User,
+    utctime: number,
```
Comment (Mateiadrielrafael) [inline index.ts]: again, pretty inconsistent naming, shouldn't it be utcTime?

File: index.ts [hunk 4]
```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
     return false;
 };
```
Comment (Mateiadrielrafael) [inline index.ts]: users can just be a User array and you can use for of... Object.entries(users)

Comment (Mateiadrielrafael) [reply index.ts]: You can also just rewrite this function useing Array.some

File: index.ts [hunk 5]
```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
...
+    if (arr.length === 0) {
+        return 0;
+    }
+    // for (let i = 1; i <= Math.max(...arr) + 1; i++) {
+    //     // console.log(i + "|" + arr.indexOf(+i) + "|" + Math.max(...arr));
+    //     if (arr.indexOf(+i) === -1) {
+    //         // console.log("done");
+    //         return +i;
+    //     }
+    // }
```
Comment (Mateiadrielrafael) [inline index.ts]: Again, stop having ghost code

File: index.ts [hunk 6]
```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
...
     wss.clients.forEach(function each(ws) {
         ws.send(JSON.stringify([category, data]));
     })
 };
 
+const updateMessagesFromDb = async () => {
+    console.log("Getting messages from database...");
+    const rows = await knex.from("messages").select("*");
+    let messageIds: Array<number> = [];
+    for (let message of rows) {
```
Comment (Mateiadrielrafael) [inline index.ts]: you can use const in for of loops

File: index.ts [hunk 7]
```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
...
+        let userFound = false;
+        let msgUser: any;
```
Comment (Mateiadrielrafael) [inline index.ts]: why this any?

File: index.ts [hunk 8]
```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
...
+    return Math.max(...arr) + 1;
+};
+
 const sendToClients = (category, data) => {
     wss.clients.forEach(function each(ws) {
         ws.send(JSON.stringify([category, data]));
     })
 };
 
+const updateMessagesFromDb = async () => {
```
Comment (Mateiadrielrafael) [inline index.ts]: ngl this function is super duper messy

File: index.ts [hunk 9]
```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
...
+            // console.log(channel);
+            channels[channel.id] = {
+                name: channel.name,
+                id: channel.id,
+                messages: channel.messages
+            };
+        }
+    } catch (err) {
+        console.error(err);
+        throw err;
```
Comment (Mateiadrielrafael) [inline index.ts]: Doesnt rethroing it do the same thing?

File: index.ts [hunk 10]
```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
...
+    }
+};
+
+const deleteOldUsers = (sNum) => {
+    console.log(sNum + ": Deleting old users...");
```
Comment (Mateiadrielrafael) [inline index.ts]: use template strings

File: index.ts [hunk 11]
```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
...
+    try {
+        // knex("users")
+        //     .where("snum", "<", sNum)
+        //     .del();
+        knex.raw('delete from users where snum <= ' + sNum + ";")
+            .catch(err => {
+                console.error(err);
+                throw err;
+            });
```
Comment (Mateiadrielrafael) [inline index.ts]: why use raw?

File: index.ts [hunk 12]
```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
...
+    }
+};
+
+const deleteOldUsers = (sNum) => {
+    console.log(sNum + ": Deleting old users...");
+    try {
+        // knex("users")
+        //     .where("snum", "<", sNum)
+        //     .del();
+        knex.raw('delete from users where snum <= ' + sNum + ";")
```
Comment (Mateiadrielrafael) [inline index.ts]: also, what happens if I try to sql inject yuo trough the sNum

File: index.ts [hunk 13]
```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
...
+        const sNum = rows[0].snum;
+        try {
+            await knex("serverid").update({snum: Number(Number(sNum) + 1)});
+        } catch (err) {
+            console.error(err);
+            throw err;
+        }
+        resolve(sNum);
+    });
+};
```
Comment (Mateiadrielrafael) [inline index.ts]:
```suggestion
const incrementSNum = async () => {
    const rows = await knex("serverid");
    const sNum = rows[0].snum;
    try {
        await knex("serverid").update({snum: Number(Number(sNum) + 1)});
    } catch (err) {
        console.error(err);
        throw err;
    }
    return sNum;
};
```

File: index.ts [hunk 14]
```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
...
+
+const init = async () => {
+    console.log("Initializing...");
+    const sNum = await incrementSNum();
+    deleteOldUsers(sNum);
+    updateMessagesFromDb();
+    updateChannelsFromDb();
```
Comment (Mateiadrielrafael) [inline index.ts]: Either await all of them if you want those to happen in a sequence or await Promise.all if you want them in parallel

File: index.ts [hunk 15]
```diff
@@ -94,6 +277,13 @@ wss.on('connection', function connection(ws) {
...
                 for (let i in messages) {
                     if (messages[i].id === message) {
+                        knex("messages")
+                            .where({id: message})
+                            .del()
+                            .catch((err) => {
+                                console.error(err);
+                                throw err;
+                            });
                         messages.splice(+i, 1);
```
Comment (Mateiadrielrafael) [inline index.ts]: this +i  feels so hacky

File: index.ts [hunk 16]
```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
...
+    console.log("HTTP server listening on: " + port);
+    init();
+});
+
+const wss = new Server({server});
+
 wss.on('connection', function connection(ws) {
     ws.isAlive = true;
-    ws.on('message', function incoming(data) {
+    ws.on('message', async function incoming(data) {
```
Comment (Mateiadrielrafael) [inline index.ts]: why not an arrow function? Also this is exremly messy, eiher split this into a bunch of functions or use something like gears

File: index.ts (hunk 11 above)
Comment (xWafl) [reply index.ts]: You can't use comparative where statements in knex

File: helpers.ts [hunk 17]
```diff
@@ -0,0 +1,12 @@
+import {User} from "./interfaces";
+
+const getHighestFromArr = (arr: Array<number>) => {
+    if (arr.length === 0) {
```
Comment (Vyctor661) [inline helpers.ts]: you can do `if (!arr.length)`

Comment (Vyctor661) [reply helpers.ts]: actually you can make it even shorter by doing `return arr.lenght ? Math.max(...arr) + 1 : 0;`

File: index.ts [hunk 18]
```diff
@@ -1,135 +1,98 @@
...
 
-const checkUser = (checkedUser: User, users: Object) => {
-    for (let i in users) {
-        if (users[i].id === checkedUser.id) {
-            return true;
-        }
-    }
-    return false;
-};
+const checkUser = (checkedUser: User, users: Object) => Object.entries(users).some(l => l[1].id === checkedUser.id);
```
Comment (Vyctor661) [inline index.ts]: you already have the `checkUser` function in `helper.ts`

File: index.ts [hunk 19]
```diff
@@ -1,135 +1,98 @@
...
                     .catch((err) => {
                         console.error(err);
+                        throw err;
                     });
-                // console.log(msgInfo.user);
-                msgInfo.user.messages.push(newMessage.id);
+
+                msgInfo.user.messages.push(msgInfo.id);
                 const isUserAuth = checkUser(msgInfo.user, users);
                 if (isUserAuth === true) {
```
Comment (Vyctor661) [inline index.ts]: just do `if (isUserAuth) {`

File: helpers.ts [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+        return 0;
+    }
+    return Math.max(...arr) + 1;
+};
+
+const checkUser = (checkedUser: User, users: Object) => Object.entries(users).some( l => l[1].id === checkedUser.id);
```
Comment (Mateiadrielrafael) [inline helpers.ts]: why is users an Object?

File: index.ts [hunk 21]
```diff
@@ -1,103 +1,98 @@
...
-class Message {
-    constructor(public user: User, public utctime: string, public date: string, public message: string, public channel: string, public id: number, public active: boolean){}
-}
-
-class Channel {
-    constructor(public name: string, public id: number, public messages: Array<Message>) {}
-}
+import {knex} from './knex';
+import {User, Message, Channel} from './interfaces';
+// import {getHighestFromArr} from "./helpers";
```
Comment (Mateiadrielrafael) [inline index.ts]: stop keeping dead code

File: index.ts [hunk 22]
```diff
@@ -106,30 +101,87 @@ wss.on('connection', function connection(ws) {
                 sendToClients("channelList", channels);
             }
             if (category === "newUser") {
-                const theUser = new User (message.username, message.usernum, 1, ++highestUserId, [0]);
+                const rows = await knex("serverid");
+                const sNum = rows[0].snum;
+                const theUser: User = {
+                    username: message.username,
+                    usernum: message.userNum,
```
Comment (Mateiadrielrafael) [inline index.ts]: can you like always use `userNum` or `usernum`?

File: index.ts [hunk 23]
```diff
@@ -106,30 +101,87 @@ wss.on('connection', function connection(ws) {
...
+                    .catch(err => {
+                        console.error(err);
+                        throw err;
+                    });
+                knex("messages")
+                    .where({channel: message})
+                    .del()
+                    .catch(err => {
+                        console.error(err);
+                        throw err;
```
Comment (Mateiadrielrafael) [inline index.ts]: You have this piece of code in a lot of places, why not use await and use a big try catch block?

File: init.ts [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+        messages.push(newMessage);
+    }
+};
+
+const updateChannelsFromDb = async (channels: Object) => {
+    console.log("Getting channels from database...");
+    try {
+        const rows = await knex.from("channels").select("*");
+        for (const channel of rows) {
+            // console.log(channel);
```
Comment (Mateiadrielrafael) [inline init.ts]: PLEASE STOP KEEPING GHOST CODE!!!

File: init.ts [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+        console.error(err);
+        throw err;
+    }
+    return sNum;
+};
+
+const getHighestId = async (): Promise<number> => {
+    interface idRet {
+        id: number
+    }
```
Comment (Mateiadrielrafael) [inline init.ts]: scoped interface, really?

File: init.ts (hunk 24 above)
Comment (xWafl) [reply init.ts]: oops i missed that one

File: index.ts (hunk 22 above)
Comment (xWafl) [reply index.ts]: userNum breaks with the sql but usernum looks worse imo, i'll migrate all of em to usernum

File: helpers.ts (hunk 20 above)
Comment (xWafl) [reply helpers.ts]: because it is? that's jsut the way i did things

File: init.ts (hunk 25 above)
Comment (xWafl) [reply init.ts]: typescript was throwing an error about id not being a property of object

File: index.ts (hunk 23 above)
Comment (xWafl) [reply index.ts]: i guess

File: init.ts [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+    const ids: Array<idRet> = await knex("messages").select("id");
+    const arrIds = Array.from(ids, l => l.id);
```
[the remark this answers is not in the document]
Comment (Mateiadrielrafael) [reply init.ts]: Isn't this cleaner?

```suggestion
    const arrIds = ids.map(({ id }) => ld);
```

File: helpers.ts (hunk 20 above)
Comment (xWafl) [reply helpers.ts]: it's not

**Pull request merged**